### ISSUE TITLE
Add OptionalTypeDesc to the AST

### ISSUE THREAD

**Issue by heshanpadmasiri** [ballerina-platform/nballerina#867]
Currently we represent both T|() and T? similarly in the AST (Both as BinaryTypeDesc with right as BuiltinTypeDesc null) with only difference being for T? case BinaryTypeDesc’s right having same start and end pos. Instead we need to make this difference explicit in the AST.

**Comment by heshanpadmasiri** [comment]
@jclark for this I am thinking about adding the fallowing to the AST
```
public const OptionalTypeOp = "?";

public type OptionalTypeDesc record {|
    *PositionFields;
    OptionalTypeOp op;
    Position opPos;
    TypeDesc td;
|};
```
One thing I am not clear is spec says [`optional-type-descriptor := type-descriptor ?`](https://ballerina.io/spec/lang/2021R1/#section_5.6.9)  but our grammar says [`optional-type-desc = postfix-type-desc [ "?" ]` ](https://github.com/ballerina-platform/nballerina/blob/2998c3cc1b59d019525878c7217375b740d8652f/docs/subset13.md?plain=1#L113) (if I am not mistaken this means "?" is optional). I assume our grammar needs to be changed to match the spec as well.

**Comment by jclark** [comment]
postfix-type-desc grammar is wrong. I just fixed it in db800a78382ec24c02618399be8d78e3b1c97c8c

**Comment by jclark** [comment]
Just change UnaryTypeOp to `"!"|"?"`, and you can handle OptionalTypeDesc  with UnaryTypeDesc (which it is, since unary means has one operand).
